### ISSUE TITLE
Postgres ON CONFLICT WHERE Support

### ISSUE THREAD

**Issue by austinarbor** [komapper/komapper#1414]
### Use case

In PostgreSQL, there is an additional `WHERE` clause that needs to be included for `ON CONFLICT` if the unique index is a partial index.

```SQL
CREATE TABLE names(
    id BIGINT PRIMARY KEY NOT NULL,
    first_name VARCHAR(255),
    last_name VARCHAR(255),
    deleted_at TIMESTAMP WITH TIME ZONE NULL
);

CREATE UNIQUE INDEX idx_uq_last_name ON names(last_name) WHERE deleted_at IS NULL;

-- success
INSERT INTO names(id, first_name, last_name, deleted_at) values(1, 'austin', 'something' ,null);
-- ERROR: duplicate key value violates unique constraint "idx_uq_last_name" Detail: Key (last_name)=(something) already exists.
INSERT INTO names(id, first_name, last_name, deleted_at) values(2, 'austin2', 'something', null); 
-- ERROR: there is no unique or exclusion constraint matching the ON CONFLICT specification
INSERT INTO names(id, first_name, last_name, deleted_at) values(2,'austin2', 'something', null)
    ON CONFLICT (last_name) DO UPDATE SET first_name = excluded.first_name; 
-- ERROR: there is no unique or exclusion constraint matching the ON CONFLICT specification
-- this is what komapper generates when you have QueryDsl.insert(a).onDuplicateKeyUpdate(prop).where { ... }
INSERT INTO names as t (id, first_name, last_name, deleted_at) values(2,'austin2', 'something', null)
    ON CONFLICT (last_name) DO UPDATE SET first_name = excluded.first_name WHERE t.deleted_at is null; 
-- success
-- note the additional WHERE clause immediately after ON CONFLICT which is needed for partial indexes
INSERT INTO names as t (id, first_name, last_name, deleted_at) values(2,'austin2', 'something', null)
    ON CONFLICT (last_name) WHERE t.deleted_at IS NULL DO UPDATE SET first_name = excluded.first_name WHERE t.first_name='austin';
```

### Describe the solution you'd like
A way to provide the `ON CONFLICT WHERE` clause

### Versions

- Komapper: 3.1.0
- Kotlin: 2.0.21
- JVM: 21
- JDBC Driver or R2DBC Driver: JDBC
- Database: Postgresql
- OS: OSX

**Comment by nakamura-to** [comment]
Thank you for your feedback.

To differentiate it from the existing `where`, we need to decide on a keyword in Komapper's DSL that corresponds to `ON CONFLICT WHERE`. For example, based on PostgreSQL documentation, `indexPredicate` could be a potential candidate.

```kotlin
val n = Meta.name
val name = ...
val query = QueryDsl.insert(n)
    .onDuplicateKeyUpdate(n.lastName)
    .indexPredicate {
        n.deletedAt.isNull() 
    }
    .set { excluded ->
        n.firstName eq excluded.firstName
    }.where {
        n.firstName eq "austin"
    }.single(name)
```

Do you have any suggestions for a better name? I would love to hear your thoughts.

**Comment by nakamura-to** [comment]
@austinarbor  I’ve implemented this in #1433. 
Please feel free to share any feedback.

**Comment by austinarbor** [comment]
@nakamura-to this looks great! i tested locally and the newly generated sql is perfect. thank you!
